File: 18-week-14/01-day-3-to-5--validate-bst/solutions/validBTS.js
class Node {
  constructor(val, left = null, right = null){
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

const isValid = (root, low, high) => {
  if(!root) {
      return true;
  }
  if(root.val <= low || root.val >= high) {
      return false;
  }
  return (isValid(root.left, low, root.val) && isValid(root.right, root.val, high));
  
}

function isValidBST(root) {
  return isValid(root, -Infinity, Infinity);
}

if (require.main === module) {

// add your own tests in here
const T = (v, l, r) => new Node(v, l, r)

console.log("Expecting: true");
console.log("=>", isValidBST(T(2, T(1), T(3))));

}

module.exports = {
  isValidBST,
  Node
}

File: 18-week-14/01-day-3-to-5--validate-bst/solutions/validBTS.rb
class Node
    attr_accessor :val, :left, :right
    def initialize(val = 0, left = nil, right = nil)
        @val = val
        @left = left
        @right = right
    end
end

def isValid(root, low, high)
    !root ||
    low < root.val &&
    root.val < high &&
    isValid(root.left, low, root.val) &&
    isValid(root.right, root.val, high)
end

def is_valid_bst(root)
    return isValid(root, -Float::INFINITY, Float::INFINITY)
end

if __FILE__ == $PROGRAM_NAME

    puts "Expecting: true"
    def T(v,l,r)
        return Node.new(v, l, r)
    end
    puts "=>", is_valid_bst(T(2, T(1, nil, nil), T(3, nil, nil)))

end
